Add logfile actuator

tscrypter opened the pull request:
## Description

This PR adds a [Logfile actuator](https://docs.spring.io/spring-boot/api/rest/actuator/logfile.html). This endpoint will expose the contents of an logfile when it's configured. To log to a file, the application will need to make use of [Third Party Logging providers](https://learn.microsoft.com/en-us/dotnet/core/extensions/logging-providers#third-party-logging-providers).

The Third-party logging provider is responsible for log rotation.

This actuator can be configured with settings prefixed with `Management:Endpoints:Logfile`:
| Key | Description | Default |
|-------|------------------|-------------|
| Enabled | Whether the endpoint is enabled | true |
| ID | The unique ID of the endpoint | logfile |
| Path | The relative path at which the endpoint is exposed | logfile |
| RequiredPermissions | Permissions required to access the endpoint, when running on Cloud Foundry | Restricted |
| AllowedVerbs | An array of HTTP verbs the endpoint is exposed at. | GET |
| FilePath | The path to the log file that will be exposed. If relative, will be relative to the applications working directory. |  |


Fixes #1496  

## Quality checklist
- [ ] Your code complies with our [Coding Style](https://github.com/SteeltoeOSS/.github/blob/main/contributing-docs/contributing-code-style.md).
- [X] You've updated unit and/or integration tests for your change, where applicable.
- [ ] You've updated documentation for your change, where applicable.
      If your change affects other repositories, such as [Documentation](https://github.com/SteeltoeOSS/Documentation), [Samples](https://github.com/SteeltoeOSS/Samples) and/or [MainSite](https://github.com/SteeltoeOSS/MainSite), add linked PRs here.
- [X] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.
- [X] You've added required license files and/or file headers (explaining where the code came from with proper attribution), where code is copied from StackOverflow, a blog, or OSS.

tscrypter commented:
I'll update this PR with a link to a documentation update PR. I also wanted to check if the [Management samples](https://github.com/SteeltoeOSS/Samples/tree/main/Management/src) would be the correct one to update.

TimHess commented:
Hi @tscrypter,
Thanks for getting this going, its looking like a good start. Some notes:
- Continue to ignore any failure with the `Steeltoe Sonar Analysis` check (if it happens again), we're working on that separately
- Can you change it to "Log File" and `LogFile` instead of "Logfile" or `Logfile` in code and comments (the path `/actuator/logfile` is fine though)
- Do you have any concerns with using [`File.ReadAllTextAsync`](https://github.com/SteeltoeOSS/Steeltoe/pull/1499/files#diff-58cf30762908956733c129bce8dde3c40c35b3488fe021bd0a4b13d61b9592ceR37) on larger text files? I don't know what the best option will be, but this also doesn't seem particularly compatible with reading a range of the file

tscrypter commented:
> Hi @tscrypter, Thanks for getting this going, its looking like a good start. Some notes:
> 
> * Continue to ignore any failure with the `Steeltoe Sonar Analysis` check (if it happens again), we're working on that separately
> * Can you change it to "Log File" and `LogFile` instead of "Logfile" or `Logfile` in code and comments (the path `/actuator/logfile` is fine though)
> * Do you have any concerns with using [`File.ReadAllTextAsync`](https://github.com/SteeltoeOSS/Steeltoe/pull/1499/files#diff-58cf30762908956733c129bce8dde3c40c35b3488fe021bd0a4b13d61b9592ceR37) on larger text files? I don't know what the best option will be, but this also doesn't seem particularly compatible with reading a range of the file

Hi @TimHess ,

Of course, I can make those updates tonight.

Good catch on the [`File.ReadAllTextAsync`](https://github.com/SteeltoeOSS/Steeltoe/pull/1499/files#diff-58cf30762908956733c129bce8dde3c40c35b3488fe021bd0a4b13d61b9592ceR37). I can update this to read bytes while I implement the range header support.

Regarding handling large log file sizes, maybe introducing a `MaxBytes` config setting. If not set, the endpoint can retrieve the entire file; else, return no more than `MaxBytes`. 

Assuming 2048 byte log file:
| MaxBytes | Range Header | Data returned|
|----------------|----------------------|------------------------|
|                    |                           |  2048 bytes         |
|  1024         |                           |  first 1024 bytes |
|                    | bytes=0-1024 |  first 1024 bytes |
| 1024          | bytes=1000-2048 | bytes 1000-2024 |

bart-vmware commented on src/Management/src/Endpoint/Actuators/LogFile/LogFileEndpointHandler.cs:
```diff
@@ -0,0 +1,57 @@
+        _logger = loggerFactory.CreateLogger<LogFileEndpointHandler>();
+    }
+
+    public async Task<string> InvokeAsync(object? argument, CancellationToken cancellationToken)
```
If this returned `Task<string?>`, the middleware could respond with 404 when `null` is returned by overriding `WriteResponseAsync`. Should catch a `FileNotFoundException` here to detect this. This is normal behavior; some logging frameworks don't create the file until there's something to log, so it shouldn't log a warning. But when the configured path is `null`, the endpoint was turned on without configuring a path, which is a mistake, so a warning should be logged in that case (and `null` should be returned).

bart-vmware reviewed (changes requested):
The Sonar issue has been fixed, please merge/rebase this PR on main. `TestContext` was renamed to `SteeltoeTestContext`, please adapt tests accordingly to fix build errors. In tests, async methods that take a `CancellationToken` (such as `app.StartAsync` and `httpClient.GetAsync`) should be passed `TestContext.Current.CancellationToken`, this is to prepare for xUnit3.

To fix PublicAPI warnings, it's easiest to clear `PublicAPI.Unshipped.txt`, close the file, and then use the roslyn fixer (project-wide) to recreate the file.

tscrypter commented on src/Management/src/Endpoint/Actuators/LogFile/ConfigureLogFileEndpointOptions.cs:
```diff
@@ -0,0 +1,26 @@
+        ArgumentNullException.ThrowIfNull(configuration);
+    }
+
+    public override void Configure(LogFileEndpointOptions options)
+    {
+        ArgumentNullException.ThrowIfNull(options);
+
+        base.Configure(options);
+    }
```
Addressed in 24c17c303dd1ce5b3bdc492681cc17d02e299911

bart-vmware replied:
The change is good, but it looks like that commit isn't part of this PR.

bart-vmware commented on src/Management/src/Endpoint/Actuators/LogFile/LogFileEndpointHandler.cs:
```diff
@@ -0,0 +1,56 @@
+
+    public async Task<string> InvokeAsync(object? argument, CancellationToken cancellationToken)
+    {
+        _logger.LogTrace("Invoking LogfileEndpointHandler with argument: {Argument}", argument);
```
```suggestion
        _logger.LogTrace("Invoking {Handler} with argument: {Argument}", nameof(LogFileEndpointHandler), argument);
```
Was still using the old casing.

tscrypter commented:
> Thanks for addressing the feedback. I wrongly commented about clearing `PublicAPI.Unshipped.txt`. The first line (`#nullable enable`) needs to stay, and then the analyzer can refill it.
> 
> Do you need some tips regarding ranges and file I/O, or did you just not get to it yet?

Thanks @bart-vmware for the guidance. I haven't gotten to it yet, I will likely have some time this evening or tomorrow. I would like your input on the potential implementation before I refactor the code to support it.

I plan on introducing `LogFileEndpointRequest` and `LogFileEndpointResponse` classes for the additional range parameters. I'll refactor the `ILogFileEndpointHandler` interface with these and get the tests that are currently in place back to working.

For the `LogFileEndpointMiddleware`, I know I'll have to override the `ParseRequestAsync` to read the original HttpRequest headers, but I *think* I'll also have to override the `WriteResponseAsync` (to put the correct response headers). I also don't think I can use the base `EndpointMiddleware.WriteResponseAsync` because the `LogFileEndpointResponse` shouldn't be JsonSerialized.

bart-vmware commented:
> > Thanks for addressing the feedback. I wrongly commented about clearing `PublicAPI.Unshipped.txt`. The first line (`#nullable enable`) needs to stay, and then the analyzer can refill it.
> > Do you need some tips regarding ranges and file I/O, or did you just not get to it yet?
> 
> Thanks @bart-vmware for the guidance. I haven't gotten to it yet, I will likely have some time this evening or tomorrow. I would like your input on the potential implementation before I refactor the code to support it.
> 
> I plan on introducing `LogFileEndpointRequest` and `LogFileEndpointResponse` classes for the additional range parameters. I'll refactor the `ILogFileEndpointHandler` interface with these and get the tests that are currently in place back to working.
> 
> For the `LogFileEndpointMiddleware`, I know I'll have to override the `ParseRequestAsync` to read the original HttpRequest headers, but I _think_ I'll also have to override the `WriteResponseAsync` (to put the correct response headers). I also don't think I can use the base `EndpointMiddleware.WriteResponseAsync` because the `LogFileEndpointResponse` shouldn't be JsonSerialized.

Yes, that sounds about right. Some additional thoughts:

- Perhaps make the endpoint return `IAsyncEnumerable<byte[]>` (4-16k buffer) to stream large files without fully loading them into memory first? Good for scalability, but kinda conflicts with the next bullet.
- What's a bit odd is that Spring returns `Content-Type: text/plain;charset=UTF-8`, but with a range it may not be proper UTF-8 at all (thinking about BOM, multi-byte chars and surrogate pairs).
- Should return HTTP 416 on multi-range requests, see https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Range
- We don't want to lock while reading a file, potentially preventing the logger from writing to it. So should probably open with `FileShare.ReadWrite`, but I haven't tested any of that.
- May want to use HTTP compression if the client supports it, don't know if this is feasible

bart-vmware reviewed (commented):
Nice progress, I've added a few remarks.

bart-vmware commented:
/azp run Steeltoe.All

bart-vmware commented:
@tscrypter Are you still working on this?

tscrypter commented:
> @tscrypter Are you still working on this?

Hi @bart-vmware, yes I am. Sorry for the delay, some critical issues came up at work. I'm hoping to have this wrapped up this weekend.
